### PR TITLE
fix: prevent propagating IO errors when serving PWA icons

### LINKED ISSUE #17643: PwaIcon: Failed to store the icon image into the stream provided
### Description of the bug

When an exception occurs while [writing a PWA icon](https://github.com/vaadin/flow/blob/73a08b0805365b17218dccb5501d091d4b800834/flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java#L238) this exception is thrown uncontrolled until it's catched by a global error handler (or worst case: never).

This is happening quite often on our systems. The underlying error is mostly `` java.io.IOException: Broken pipe`` which happens e.g. when quickly reloading the site or aborting the site load.

<details><summary>Logs</summary>

```
2023-09-18 12:32:43 WARN  [http-nio-8080-exec-1] [omponents.error.ErrorUtil] Unhandled Exception detected!
2023-09-18 12:32:43 ERROR [http-nio-8080-exec-1] [omponents.error.ErrorUtil] Fatal Error
java.io.UncheckedIOException: Failed to store the icon image into the stream provided
        at com.vaadin.flow.server.PwaIcon.write(PwaIcon.java:240) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.communication.PwaHandler.lambda$init$1bc33a08$1(PwaHandler.java:83) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.communication.PwaHandler.handleRequest(PwaHandler.java:153) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1529) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:398) ~[flow-server-24.1.11.jar!/:24.1.11]
        at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:106) ~[vaadin-spring-24.1.11.jar!/:?]
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:205) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:642) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:408) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:313) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:277) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:141) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:178) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1081) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:974) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1011) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.0.11.jar!/:6.0.11]
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:205) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.security.web.ObservationFilterChainDecorator$FilterObservation$SimpleFilterObservation.lambda$wrap$1(ObservationFilterChainDecorator.java:479) ~[spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.security.web.ObservationFilterChainDecorator.lambda$wrapUnsecured$1(ObservationFilterChainDecorator.java:90) ~[spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:219) [spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:191) [spring-security-web-6.1.3.jar!/:6.1.3]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:352) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:268) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:109) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) [spring-web-6.0.11.jar!/:6.0.11]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) [spring-web-6.0.11.jar!/:6.0.11]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:174) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:149) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:166) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:341) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:391) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:894) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1740) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) [tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-10.1.12.jar!/:?]
        at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: org.apache.catalina.connector.ClientAbortException: java.io.IOException: Broken pipe
        at org.apache.catalina.connector.OutputBuffer.realWriteBytes(OutputBuffer.java:347) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.appendByteArray(OutputBuffer.java:750) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.append(OutputBuffer.java:677) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.writeBytes(OutputBuffer.java:382) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:360) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:97) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:90) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at com.vaadin.flow.server.PwaIcon.write(PwaIcon.java:238) ~[flow-server-24.1.11.jar!/:24.1.11]
        ... 67 more
Caused by: java.io.IOException: Broken pipe
        at sun.nio.ch.FileDispatcherImpl.write0(Native Method) ~[?:?]
        at sun.nio.ch.SocketDispatcher.write(Unknown Source) ~[?:?]
        at sun.nio.ch.IOUtil.writeFromNativeBuffer(Unknown Source) ~[?:?]
        at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
        at sun.nio.ch.IOUtil.write(Unknown Source) ~[?:?]
        at sun.nio.ch.SocketChannelImpl.write(Unknown Source) ~[?:?]
        at org.apache.tomcat.util.net.NioChannel.write(NioChannel.java:118) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.NioEndpoint$NioSocketWrapper.doWrite(NioEndpoint.java:1381) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketWrapperBase.doWrite(SocketWrapperBase.java:764) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketWrapperBase.writeBlocking(SocketWrapperBase.java:589) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.tomcat.util.net.SocketWrapperBase.write(SocketWrapperBase.java:533) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.Http11OutputBuffer$SocketOutputBuffer.doWrite(Http11OutputBuffer.java:540) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.filters.ChunkedOutputFilter.doWrite(ChunkedOutputFilter.java:112) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.http11.Http11OutputBuffer.doWrite(Http11OutputBuffer.java:193) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.coyote.Response.doWrite(Response.java:616) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.realWriteBytes(OutputBuffer.java:334) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.appendByteArray(OutputBuffer.java:750) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.append(OutputBuffer.java:677) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.writeBytes(OutputBuffer.java:382) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.OutputBuffer.write(OutputBuffer.java:360) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:97) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at org.apache.catalina.connector.CoyoteOutputStream.write(CoyoteOutputStream.java:90) ~[tomcat-embed-core-10.1.12.jar!/:?]
        at com.vaadin.flow.server.PwaIcon.write(PwaIcon.java:238) ~[flow-server-24.1.11.jar!/:24.1.11]
        ... 67 more
```

</details>

### Expected behavior

The "error" can neither be fixed nor handled and causes a lot of log spamming, so it would be nice if you could catch it beforehand and log it to e.g. debug level.

### Minimal reproducible example

This exception is hard to reproduce, as it only happens e.g. when the client cuts the connection, which is not easily controllable.

To reproduce the problem I recommend downloading the Vaadin starter, adding PWA support and hitting the refresh button very often.

### Versions

- Vaadin / Flow version: 24.1.9 (currently latest version)

### PR DESCRIPTION
**Pull request by mcollovati**
## Description

An IOException when writing a PWA icon on the HTTP response can neither be fixed nor handled and causes a lot of log spamming. This change catches the error and logs it at debug level.

Fixes #17643

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java [hunk 1]
```diff
@@ -81,6 +84,9 @@ private void init(PwaRegistry pwaRegistry) {
                         }
                         try (OutputStream out = response.getOutputStream()) {
                             icon.write(out);
+                        } catch (UncheckedIOException ex) {
+                            LoggerFactory.getLogger(PwaHandler.class)
+                                    .debug("Error serving PWA image", ex);
```
Comment (czp13) [inline flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java:89]: Maybe?
```suggestion
                                    .debug("Error serving PWA icon", ex);
```
just to be more consistant with the namings.

File: flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java [hunk 2]
```diff
@@ -93,4 +98,40 @@ public void handleRequest_pwaRegistryConfigIsEnabled_handlerIsInitializedOnce()
 
         Mockito.verify(registry, Mockito.times(1)).getIcons();
     }
+
```
Comment (czp13) [inline flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java:101]: If you are this class maybe:
```
    private VaadinSession session = Mockito.mock(VaadinSession.class);
    private VaadinRequest request = Mockito.mock(VaadinRequest.class);
    private VaadinResponse response = Mockito.mock(VaadinResponse.class);
```
These could be marked finals :), as they only get value once.

**Review by czp13 — COMMENTED** [review]
Added some smallish comments, and the code change and the created test seem good to me! 🙇

File: flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java (hunk 2 above)
Comment (mcollovati) [reply flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java:101]: done

File: flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java (hunk 1 above)
Comment (mcollovati) [reply flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java]: done

Comment (czp13) [reply flow-server/src/main/java/com/vaadin/flow/server/communication/PwaHandler.java]: Thank you vm! 😊

File: flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java (hunk 2 above)
Comment (czp13) [reply flow-server/src/test/java/com/vaadin/flow/server/communication/PwaHandlerTest.java:101]: Thank you for this one too! 🙇

**Review by czp13 — APPROVED** [review]
Thank you Marco for the work, this is a good one! 👍 
(especially thanks for tackling the review comments)

I am happy to approve, but if you feel needing another pair of eyes, we can wait for another review as well :).

**Pull request merged**